File: 530. Minimum Absolute Difference in BST/530. Minimum Absolute Difference in BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.min_diff = float("inf")
        self.prev = None

        def dfs(node):
            if not node or self.min_diff == 1:
                return

            dfs(node.left)

            if self.prev:
                self.min_diff = min(self.min_diff, abs(self.prev - node.val))

            self.prev = node.val

            dfs(node.right)

        dfs(root)
        return self.min_diff

File: 530. Minimum Absolute Difference in BST/530. Minimum Absolute Difference in BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  int minDiff = Integer.MAX_VALUE;
  TreeNode prev = null;

  public int getMinimumDifference(TreeNode root) {
    dfs(root);
    return minDiff;
  }

  private void dfs(TreeNode node) {
    if (node == null || minDiff == 1) {
      return;
    }

    dfs(node.left);

    if (prev != null) {
      minDiff = Math.min(minDiff, Math.abs(prev.val - node.val));
    }

    prev = node;

    dfs(node.right);
  }
}
